### ISSUE TITLE
ability to use fixtures/factories in dev/tests

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#46]
whichever you prefer. shouldn't be too difficult.

**Comment by seawatts** [comment]
What does this mean? Like in unit tests?

**Comment by samselikoff** [comment]
Currently fixtures work in dev (actually in tests too, if you don't have any factories). You only use factories in tests. I want to have a clear path if, e.g. you only want to use factories. So you've set them up for your tests, and you want to be able to set up a default scenario for development too, using your factories to generate some data.

**Comment by seawatts** [comment]
That makes sense. Is there any plan to be able to use this for unit tests?

**Comment by samselikoff** [comment]
ohhh using mirage in unit tests. Hm, hadn't thought about that, what unit test would you need to mock out HTTP requests, your adapter or something?

**Comment by seawatts** [comment]
Lol actually that is exactly what I was talking about. But I could also see it useful just to use the fixtures for other unit tests as well. Not necessarily mocking the http requests, but since we already have the fixtures defined, I would rather not have to define them again.

**Comment by samselikoff** [comment]
hm what would you use the fixtures for, if not http mocks?

**Comment by varblob** [comment]
I'm actually thinking about this at this very moment.  I would really like to use the factories I created to  generate ember data models with default values.  I guess I could use something like https://github.com/danielspaniel/ember-data-factory-guy, but I would like to avoid having to create factories twice.

**Comment by samselikoff** [comment]
yes. The problem is, Mirage is really about http mocks. It actually has nothing (very little) to do with ember, it's just using Ember CLI to wire up the server. I think it's important to keep it that way because core Mirage could very easily be extracted and used e.g. in a react app.

I could be persuaded otherwise, but right now I don't think Mirage has any business creating ember data models. I know it's a drag and perhaps there's room for another layer that will simplify this, but mirage in testing is really about "given this data in the server, here's what my ember app does". The factories are about simply reproducing your API responses. I know you guys know this but again, my current thinking is using the factories for something like making ember data models is out of scope.

**Comment by varblob** [comment]
That makes sense to me to focus on the core of pretending to be a server first.

Issues like generating ember-models could exist in a separate addon or perhaps core would be pushed to a seperate repo and ember bits would remain in this repo?  It would be great to also have some code generation from server languages to mirage especially given that mirage is moving towards having schema/validations/serializer.  I pretty much have factories, serializers, routes etc all specified on my server side anyways.  

Until that day I ended up using this method to avoid maintaining double factories.

     import MirageInitializer from 'manage-app/initializers/ember-cli-mirage'
     beforeEach: function(){
       MirageInitializer.initialize()
     }

and then I basically wrap the server call in a function that pushes the outcome into the store.... don't hate me.

**Comment by seawatts** [comment]
So what I was thinking is that you don't have to use server in unit tests but just have the ability to use the factories raw. 

So it doesn't have to anything to so with ember or ember data. Just inject factories into the unit tests and then they will have access to the outputed json object and the user can then create a ember data model on their own with that json if they want.

**Comment by varblob** [comment]
@seawatts Yah that's more or less what I was going for, this was just the fastest way I found to do it.  If you find a nicer tie in point I would be very keen to try it.

**Comment by samselikoff** [comment]
I see..interesting. @varblob haha I don't hate you, necessity is the mother of invention :)

I could *definitely* see a future where the factory layer is separate, and Mirage just happens to use those factories, as well as another layer that uses them to create ED models. But (just thinking out loud) the relationships may bite us. I'll have to see how it plays out once I add relationships support. Because once schema/models land, factories will be about creating Mirage models, no longer just raw JSON. And it has to be like that for relationships to be supported. (This is similar to factory_girl, which creates AR models).

We'll see what happens but, I think we should focus on an ED factory lib, like the one that already exists. There will definitely be some duplication but, also, in a lot of cases my ED models are different from the models returned from my API, there's not typically a one-to-one mapping. For an ED Factory layer I'd want more control over creating those specific models, that have nothing to do with my API.

Anyway for now this is a separate project, but good to keep the discussion moving + see what you guys come up with.

**Comment by eccegordo** [comment]
>  You only use factories in tests. I want to have a clear path if, e.g. you only want to use factories.

This statement should be made more obvious in the README. 

It seems logical to be able to use fixtures in tests. But perhaps not. Maybe the README or docs can have a brief statement of why this is not supported and/or is a bad idea.

**Comment by samselikoff** [comment]
@eccegordo yeah sorry, the [acceptance testing](https://github.com/samselikoff/ember-cli-mirage#acceptance-testing) section explains why fixtures aren't loaded during testing. My current position is, factories should be preferred to fixtures in tests because they let you define the state right from the test. (The line you quoted was actually a typo, it was supposed to say, "You *can* only use factories in tests", meaning currently, you can't use factories to generate development data).

The problem is right now, factories don't have relationship support. So if people take a long time to set up fixtures for dev with all the related ids, I want them to be able to use them in testing (which they can, if the factories folder doesn't exist). Once relationship support lands, factories will be the encouraged path.

I think it will be fine to still support fixtures for both dev/testing, but the relationship support in factories will make everything much easier.

**Comment by kushdilip** [comment]
@samselikoff I don't if someone else is using fixtures in tests or not but for me it's not working anymore after I upgraded form version 0.0.26 to 0.1.2 onwards. It was working earlier and I don't have factories folder at all. 
Now after upgrades all acceptance tests are failing because db is not seeded during tests. I creating factories but fixtures were far too easy to setup.

**Comment by samselikoff** [comment]
This is a bug, I'll look into it

**Comment by samselikoff** [comment]
@kushdilip could you try 0.1.5?

**Comment by danielspaniel** [comment]
It's funny because I was thinking about this very same topic for a while, and I think it would be pretty nice to reuse the factories from both our projects. I don't mind discussing some way to get that ball rolling.
 I know I would like to try out mirage, but I don't want to rewrite all my factories either, and I also want to be able to do unit tests with factories .. it's kind of a good topic.

**Comment by samselikoff** [comment]
Interesting. The biggest hurdle for me right now is the orm layer. This is where factory_girl ascertains its association information - and I'm assuming this is how factory-guy gets it's association information as well, from ED's orm.

Once I have an orm I'll be able to write factories that delegate to it, making them much simpler. But I'm certainly open to collaboration. The most reusable pieces I see are things like traits, sequences, inline factories. If factory-guy's factory layer contained these but delegated to some `schema` object for actually generating the ids and foreign keys, this might be something we could share.

**Comment by danielspaniel** [comment]
Yes, that is what stops both our projects from being factory_girl.
  Ember does not have the same ActiveModel concept that Rails has where the relationship definitions can be scanned for whatever adapter might be in use underneath. 
 I think that might happen one day.  When you say "once you have an orm", are you saying you are working on that or others are?    
 For now, I am targeting the biggest data modeling layer in ember which is ember-data, because I need that relationship info and it's the big fish.
 I also need the adapter type as well, since ActiveModel adapter needs different json than JSONAPI adapter etc .. which is another thing I would not have to worry about with an orm layer.  

 So, I think we are kinda stuck on that issue but we could let's say today share the factory definitions themselves but I would have to see what your relationship definitions are going to look like to see for sure. But once I saw that I could make it generic and try as you suggest to make it more plug and play for your project by adding overridable hooks to add id's or whatever. 
 Not sure if it's worth all the effort?  
 But I think it would be possible. And it makes for interesting pondering.

**Comment by samselikoff** [comment]
Yep I'm working on an orm for mirage, for precisely the reasons you mentioned. [Here's the branch](https://github.com/samselikoff/ember-cli-mirage/tree/models-and-more/addon/orm/associations) and the [tests](https://github.com/samselikoff/ember-cli-mirage/tree/models-and-more/tests/integration/schema). I think the first two associations are done, I'm going to hopefully have some time this weekend to work on integrating with mirage.

Route handlers will look something like this

```js
this.get('/contacts', function(schema, request) {
  return schema.contact.all();
});
```

That will return an array of Contact models, which then enables us to add a serializer layer (we could lookup a `ContactSerializer`). It also lets us have users write something simple defining their relationships

```js
// models/contact.js
export default Mirage.Model.extend({
  addresses: Mireage.hasMany()
});

// models/address.js
export default Mirage.Model.extend({
  contact: Mireage.belongsTo()
});
```

which will allow Mirage's db to set up the appropriate foreign keys. We can then use these in route handlers (`contact.addresses`) and also make factories support relationships.

**Comment by danielspaniel** [comment]
Well, this answers my question .. about how you were going to do relationships.  This is a really interesting way to go, though it does make my normally perky chest hairs go alittle limp. 
 If I get this right, you are making mirage a general ( one girdle fits all ) library by asking people to set up the orm mapping and also serializer definitions up front.  
 I am really excited to see what happens with both our libraries and how things evolve, because this discussion has given me the idea that I needed to explore the opposite ( introspection ) route alittle further.  I think what I will do is reduce the coupling to ember-data, and add a layer on top of factory guy
that can introspect an ember-data, ember-orbit or whatever to find models, relationships, the store .. and then with that layer I can call out and get the data I need for whatever new library I want to support. 
 
With your style, the user does the work up front to setup the information, and mine would be more plug and play, but I would have to write connectors for any new library I wanted to support and you would not. 

 So, I guess we will see if developers mind having to write the relationships and serialization defs twice. 
 If I were me ( and I am pretty sure I am ) I would feel like I had to stay an extra week in July to make up for all the snow days, if I had to do all that work up front to use the factories, but I think alot of people will also not mind being in school for the extra few days in the summer :)

**Comment by samselikoff** [comment]
@danielspaniel Yes you're exactly right. I toyed with the idea of trying to get relationship info from the Ember model layer (e.g. ED), but ultimately decided that this wouldn't truly be testing your adapter layer - which is really the whole point of using something like pretender or mirage to begin with. Another big reason is, Mirage really has nothing to do with Ember. Right now it's an addon since that makes the wiring easy, but I'd like to extract the core into a MIrage.js itself which could be used with other apps, like a React app. I really just want it to be a library that makes it simple to stub out your backend.

I'm hoping the extra code will ultimately keep the mirage server definition sparse enough that it won't feel like maintaining two servers. I know in the projects I'm currently working on, the mirage code helps more than it hurts, and the orm will only improve things. However I do like the idea of getting relationship info from the model layer. I'm just not sure this could be done in a general way - for example, if someone's using a custom serializer, how could you replicate the json payload?

**Comment by danielspaniel** [comment]
I think your right too @samselikoff, there is no general way to get that relationship info.  It all depends on the model implementation. Luckily in ember land, the biggest kahuna is ember-data, so it made sense to tie in to it to start out.  The benefit has been enormous, since making models with relationships with factory guy is so easy. The thing is I have to replicate the json just like you do, since I have integration test helpers that take the factory guy fixture json and pass it off to mockjax which hits whatever adapter is in use by ED.  Which is why I was so puzzled about your project .. since that json is so adapter specific. ( How the heck is Sam going to figure out what json to produce? )  Once again, you can't generalize that, you have to know what adapter they are using.  I generate jsonapi json for that adapter and regular json for the others.  With serializers, I hook into ember-data to serialize a payload so it will use whatever the serializer was setup by the user for that model.   Cause it can be really hairy to duplicate that (polymophic models and or relationships, and complex serializer hokus pokus.) 

 The thought of trying to replicate all that gave me flashbacks of the days before rails was king of startup land, and people had to drag around xml def describing all their models + relationships + serialization + blah ( rather than having it all introspected for you as rail does. )  I ran to hide in the closet and had to call my mommy to talk me out. 

  I hope your right too about making it easy to maintain two server defs, because if you pull that off I will by you your own kombucha starter kit so you can enjoy crisp and healthy mushroomy drinks all day long. You will roundly deserve it :)

**Comment by samselikoff** [comment]
Haha - indeed, it is an uncertain proposition. I got here by starting with some sugar over Pretender, then realizing we needed some solution for relationships.

I didn't realize you were generating JSON that was passed through the host application's serializer layer. I always assumed you were doing something like `store.push`. I'll have to think more about this...it certainly could be an option for us, though I'd really like mirage to work for non-ED projects too.

Perhaps we should chat about all this sometime! Are you on Slack?

**Comment by danielspaniel** [comment]
Yes, me on Slack, but not sure how to add someone. Want to talk on Saturday? I am working tonight.

**Comment by samselikoff** [comment]
Are you in the EmberJS Community team?

**Comment by danielspaniel** [comment]
What is that? I guess that means no.

**Comment by samselikoff** [comment]
https://ember-community-slackin.herokuapp.com/

**Comment by tschoartschi** [comment]
I'm using ember data factory guy right now. As I read in this thread there were thoughts to integrate/combine parts of ember data factory guy with ember-cli-mirage. Was there any progression on a collaboration between these two projects?

**Comment by samselikoff** [comment]
Unfortunately my needs relied closely on the ORM layer. It didn't seem feasible to translate factory-guy's calls into Mirage db records. So, at the moment, no, the projects are still completely separate.

**Comment by tschoartschi** [comment]
@samselikoff thanks for the quick response. Maybe I can contribute to this problem when I come to it. Right now I'm still struggeling to get Ember-CLI-Mirage up-and-running. 
@offtopic But I think it is a Pretender problem https://github.com/pretenderjs/pretender/issues/72

**Comment by danielspaniel** [comment]
Yes, very difficult to merge now with the orm layer.
What things in mirage you need that factory guy does not have?

**Comment by tschoartschi** [comment]
@danielspaniel I switched from ic-ajax to ember-ajax and they suggest to use Ember-CLI-Mirage for fixtures. Since I have a couple of none Ember Data requests I tried out Ember-CLI-Mirage. At the first glance it seems that Ember Data Factory Guy plays nicely with my setup including Ember-CLI-Mirage. I just googled around because it seems that there is overlapping functionality between Ember Data Factory Guy and Ember CLI Mirage and so I found this thread. I was just curious and I always like it if I can slim down the external dependencies of my project. Which would be the case if these two projects would contribute to each other.

**Comment by samselikoff** [comment]
If your project is already set up to use Factory Guy, I would say you should keep using it. It's totally okay to install Mirage and use it to mock out a few non-Ember Data requests. Mirage is stripped from your production build, so it won't add any weight to your final app.

**Comment by g-cassie** [comment]
Just reading through this thread. I think that there may be a good reason to add some hooks to make it possible to integrate these projects.  I started using factory-guy recently in order to support component integration tests where I want to be able to create data directly in the ember store - this is not something supported in Mirage to my knowledge and I think it is kind of outside of the scope.

Mirage still has clear use case for live development and easy acceptance tests (with factory-guy you would need to do a lot of work to setup all the mock calls for a complex route with many models and the resulting test would be more brittle than what you get with mirage).  

I need to catch up on the latest changes in Mirage but I think it would be quite useful to be able to use factory-guy api with a configurable backend.  The default backend would be ember-data store, like it is now.  A new backend would be the mirage ORM.  This way you can have one set of factory definitions and be able to use them in unit tests, integration tests, acceptance tests and development scenarios.  

I am going to continue to work with the two libraries over the next couple weeks and think a bit more on this but just wanted to share my initial thoughts and see if there is interest among @samselikoff and @danielspaniel before I start to dig into it.

**Comment by danielspaniel** [comment]
Yeah .. I have been wondering how I could get some hooks between the two going on.. but don't have a good idea yet. I know it's possible I just have not figured it out.  
 What I have been doing lately is trying to think of how to make factory guy work in development just like mirage does. But trying to dump factory guy data into mirage db actually would probably work ( but not in development .. for now ) .. since the factory guy factories are in test folder and dev environment does not load those files.  There is away around that .. just have not figured it out. 
 Anyway you are pretty spot off about the integration tests and setting up acceptance test data with factory guy. In fact with complex craziness of relationships and data, factory guy rocks the house.  I am not just saying that because I have 5 minutes to kill and have nothing else to do. If you have a case you want to show me I will prove it. There are a few tips I could show that would open your mind to the simplicity of it. 
 I was starting a contract a month ago where they were doing amazing gymnastics to stub out whacky relationships  with mirage, and I started chewing on my shoes. I just could not handle it. With factory guy it was just so easy. 
 But hey, I digress, sort of. 
 If you have any great ideas hit me ( there is factory guy channel on slack ) I am up for the chatting.

**Comment by g-cassie** [comment]
I sent you a few messages on Slack.  Granted I am not experienced enough with factory-guy at this point so am probably under estimating it.  I do really like the factory-guy api though and would love to use it exclusively for factory calls across all tests.  I will try to upgrade to the new version of Mirage soon and then have a crack at putting my proposal into some code.
